src/expr/src/scalar/like_pattern.rs

```diff
@@ -73,3 +147,774 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
+                if chars.next().is_none() {
+                    return false;
+                }
+            }
+            rest = chars.as_str();
+        }
+        if self.many {
+            // The sub-pattern might consume any number of characters, but we need to find
+            // where it terminates so we can match any subsequent sub-patterns. We do this
+            // by searching for the suffix string using str::find.
+            //
+            // We could investigate using a fancier substring search like Boyer-Moore:
+            // https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm
+            //
+            // .. but it's likely not worth it. It's slower for small strings,
+            // and doesn't really start outperforming the naive approach until
+            // haystack sizes of 1KB or greater. See benchmarking results from:
+            // https://github.com/killerswan/boyer-moore-search/blob/master/README.md
+            //
+            // Another approach that may be interesteing to look at is a
+            // hardware-optimized search:
+            // http://0x80.pl/articles/simd-strfind.html
+            if self.suffix.len() == 0 {
+                // Nothing to find... This should only happen in the last sub-pattern.
+                assert!(self.next.is_none(), "empty suffix in middle of a pattern");
+                return true;
+            }
+            // Use rfind so we perform a greedy capture, like Regex.
+            let mut found = rest.rfind(&self.suffix);
+            loop {
+                match found {
+                    None => return false,
+                    Some(offset) => {
+                        let end = offset + self.suffix.len();
+                        let tmp = &rest[end..];
+                        match &self.next {
+                            None => return tmp.len() == 0,
+                            Some(next) => {
+                                if next.is_match(tmp) {
+                                    return true;
+                                }
+                            }
+                        }
+                        // Didn't match, look for the next rfind.
+                        if offset == 0 {
+                            return false;
+                        }
+                        found = rest[..(end - 1)].rfind(&self.suffix);
+                    }
+                }
+            }
+        }
+        // No string search needed, we just use a prefix match on rest.
+        if !rest.starts_with(&self.suffix) {
+            return false;
+        }
+        rest = &rest[self.suffix.len()..];
+        match &self.next {
+            None => rest.len() == 0,
+            Some(matcher) => matcher.is_match(rest),
+        }
+    }
+}
+
+/// Breaks a LIKE pattern into a chain of sub-patterns.
+fn build_string_matcher(pattern: &str, escape: char) -> Result<StringMatcher, EvalError> {
+    let mut head = StringMatcher {
```

Comment:
You could derive the trait `Default` for `StringMatcher` and call `StringMatcher::default()` here.

Reply:
Ooh, very nice! Done.